3단계 - 질문 삭제하기 리팩터링

popotec opened the pull request:
안녕하세요, 동환님! 😊

3단계 리뷰 요청드립니다.
미션을 수행하며 질문드릴 사항이 있어서 아래 남깁니다.

[질문]

QnaServiceTest의 delete관련 테스트 메서드를 보면 answerRepository.findByQuestionAndDeletedFalse를 호출하는 부분이 있는데, QuestionService에서 question의 getAnswer을 하면 이걸 호출 하지 않아도 되어서 삭제하였습니다. (사용하지 않는 mock을 선언해놓으니 오류가 나서 지웠습니다.) 

getAnswer를 하면 .findByQuestionAndDeletedFalse를 했을때와 달리, 이미 deleted=true인 answer도 포함되어 참조되기도 하고, 테스트코드에 원래 있는 코드라 처음엔 이 방식으로 진행을 하였었는데요,

Answer entity에서 isDeleted가 true인 경우는 skip 되도록 하면 문제되지 않는다 생각하였고 객체 그래프 탐색 측면에서 repository를 이용해 가져오는 것보다 question 객체에서 가져오는게 낫지 않을까 하는 생각으로 구현하였습니다.

이렇게 구현한 것이 좋은 방향 일까요? 그리고 테스트 코드를 임의로 바꾸었는데 이것도 괜찮을지 질문드립니다. 🙏

popotec committed:
feat(DeleteHistory): 연관관계 맵핑과 Equals 수정

- Long deletedById -> User Entity와 연관관계 맵핑
- Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
fix(DeleteHistoryRepository): 데이터 중심 조회를 객체 관점으로 변경

popotec committed:
fix(User): Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
feat(Question): 연관관계 맵핑과 Equals 수정

- Long writerId -> User Entity와 연관관계 맵핑
- Answer에 대하여 일대다 양방향 연관관계 맵핑
- Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
feat(Question): 연관관계 맵핑과 Equals 수정

- Long writerId -> User Entity와 연관관계 맵핑
- Answer에 대하여 일대다 양방향 연관관계 맵핑
- Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
feat(Answer): 연관관계 맵핑과 Equals 수정

- Long writerId -> User Entity와 다대일 단방향 연관관계 맵핑
- Question에 대하여 다대일 양방향 연관관계 맵핑
- Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
feat(Answer): 연관관계 맵핑과 Equals 수정

- Long writerId -> User Entity와 다대일 단방향 연관관계 맵핑
- Question에 대하여 다대일 양방향 연관관계 맵핑
- Equals에 비즈니스 키만 포함되도록 변경

popotec committed:
refactor(QnaService): 호출 메서드명 변경된 것 반영

popotec committed:
chore: 미사용 import 삭제

popotec committed:
chore: 미사용 import 삭제

popotec committed:
fix: foreign key alias 추가

popotec committed:
fix: hashcode 개선

popotec committed:
Merge branch 'jerry92k' into step3

popotec committed:
test: 테스트 오류 수정

asserThat 이후 검증 구문 누락

popotec committed:
fix: 조회시 연관관계 객체의 식별자를 사용하지 않고 객체 자체를 사용

popotec committed:
fix: 메서드 파라미터로 받은 User와 Answer의 작성자가 같은지 확인하여 삭제 처리

popotec committed:
fix: 메서드 파라미터로 받은 User와 Question의 작성자가 같은지 확인하여 삭제 처리

- Question의 연관관계 Answer들도 탐색하며 삭제

popotec committed:
fix: answers를 빈 리스트로 초기하고 addAnswer시 관계 맵핑

popotec committed:
refactor: 삭제 요청자와 작성자가 같은지와 같은 로직을 도메인 객체로 이관

popotec committed:
refactor: AnswerRepository를 이용하여 answers를 가져오지 않고 Question의 객체 그래프 탐색으로 answers를 참조하도록 변경

- Answer: 삭제 요청자와 Answer 작석자가 달라도 이미 삭제된 것이니 오류나지 않게 삭제 유무 확인 로직 추가

popotec committed:
test: delete 예외 케이스 테스트 추가

popotec committed:
refactor: 메서드 분리

popotec committed:
docs(README.md) : 미션 정의 업데이트

na-mul commented on src/main/java/qna/domain/Question.java:
```diff
@@ -37,7 +40,7 @@
 
 	@OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
-	private List<Answer> answers;
+	private List<Answer> answers = new ArrayList<>();
```
answers도 일급콜렉션으로 바꿔볼 수 있을 것 같습니다. 그렇다면 `deleteAnswers` 메소드의 역할도 맡겨볼 수 있을 것 같아요. 

```
@Entity
public class Question extends BaseTimeEntity {

    (...)
    
    @Embedded
    private AnswerList answerList = new AnswerList();

    (...)
}

@Embeddable
public class AnswerList {

    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
    private List<Answer> answers;

    (...)
}
```

와 같은 방식으로 사용 가능하니 참고를 부탁드립니다.

na-mul commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -37,22 +37,16 @@ public Question findQuestionById(Long id) {
+    public void makeDeletedHistory(Question question) {
         List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.delete();
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.delete();
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now()));
+
+        for (Answer answer : question.getAnswers()) {
             deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
         }
         deleteHistoryService.saveAll(deleteHistories);
```
이 로직도 Service 계층에 존재하는 건 아쉬운 부분인 것 같아요. 
한번 도메인 객체 쪽으로 빼보는 건 어떨까요?
필요하다면 `DeleteHistory`의 일급콜렉션 객체를 만들어 보는 것도 하나의 예시가 될 수 있을 것 같습니다.

na-mul replied:
과제 힌트에 있는 아래 그림을 참고해보는 것도 좋을 것 같습니다.

![image](https://user-images.githubusercontent.com/58849938/141644968-ecdb22b7-d28f-48ff-bea2-1315d5b60a78.png)

na-mul commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -48,7 +48,6 @@ public void setUp() throws Exception {
     @Test
     public void delete_성공() throws Exception {
         when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
-        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
```
불필요한 mock은 제거하지 않으면 에러가 발생하지요. 
```
    @Mock
    private AnswerRepository answerRepository;
```

위 코드도 함께 지워주면 좋을 것 같아요.

na-mul replied:
더불어 `QnaService`의 `private AnswerRepository answerRepository;`도 필드와 생성자에서 제거해줄 수 있을 것 같아요.

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -88,7 +89,13 @@ public boolean isDeleted() {
 
-	public void delete() {
+	public void delete(User loginUser) throws CannotDeleteException {
+		if(isDeleted()){	/* loginUser와 다른 사용자 소유여도 이미 지워진 거면 exception 안나게 처리 */
+			return;
+		}
```
꼼꼼하게 질 해주셨네요. 

이러한 상황(`loginUser와 다른 사용자 소유여도 이미 지워진 상황`)이 발생했을 때의 처리를 어떻게 할 지는 
사람마다 의견이 분분할 수 있을 것 같습니다.
결국 이렇게 관련 로직이 모여있어야 이런 로직이 존재했었다는 것을 발견하기도 쉽고, 
수정했을 때 영향범위 또한 최소화 할 수 있을 것입니다.

객체지향적인 설계의 좋은 사례라 할 수 있겠네요 👍

na-mul reviewed (changes requested):
안녕하세요 지희님!
JPA의 마지막 3단계까지 고생이 많으셨습니다. 🙏
추가적으로 고민해보실만한 포인트들을 몇가지 코멘트로 남겨두었습니다.
본 과정에서 매우 중요한 미션인만큼, 깊게 고민해보시고 다시 request 부탁드릴께요 🙇‍♂️

na-mul commented:
> QnaServiceTest의 delete관련 테스트 메서드를 보면 answerRepository.findByQuestionAndDeletedFalse를 호출하는 부분이 있는데, QuestionService에서 question의 getAnswer을 하면 이걸 호출 하지 않아도 되어서 삭제하였습니다. (사용하지 않는 mock을 선언해놓으니 오류가 나서 지웠습니다.)

이 부분은 잘 진행하신 것 같습니다!



> getAnswer를 하면 .findByQuestionAndDeletedFalse를 했을때와 달리, 이미 deleted=true인 answer도 포함되어 참조되기도 하고, 테스트코드에 원래 있는 코드라 처음엔 이 방식으로 진행을 하였었는데요,
> 
> Answer entity에서 isDeleted가 true인 경우는 skip 되도록 하면 문제되지 않는다 생각하였고 객체 그래프 탐색 측면에서 repository를 이용해 가져오는 것보다 question 객체에서 가져오는게 낫지 않을까 하는 생각으로 구현하였습니다.
> 
> 이렇게 구현한 것이 좋은 방향 일까요? 

현재 상황에서는 좋은 방향입니다. `findByQuestionAndDeletedFalse`를 사용하게 되면 repository에 결국 도메인 로직이 포함되게 됩니다. 다만 성능을 위해서라면 도메인 객체에 모든 로직을 넣는 것을 일부 포기하는 경우가 생길 수 있겠지요. 그럴 때는 어쩔 수 없이 `findByQuestionAndDeletedFalse`등을 사용해야 할 것 같아요. 

하지만 아에 방법이 없는 것은 아닙니다. JPA에 `Specification`이라는 개념이 있는데요, 
이를 통해서 쿼리의 조건을 별도의 명세로 만들어 도메인영역에 객체화 해둘 수 있게됩니다.
JPQL을 좀 더 쉽게 쓸 수 있도록 해주는 QueryDSL등에도 비슷하게 사용할 수 있는 개념들이 있으니 참고해볼 수 있을 거에요.
다만 이러한 방법은 어플리케이션을 매우 복잡하게 만들 수 있으니, 개발팀 내에서 합의가 꼭 필요한 부분입니다.



> 그리고 테스트 코드를 임의로 바꾸었는데 이것도 괜찮을지 질문드립니다. 🙏

물론 괜찮습니다!
더 적극적으로 바꿔주셔서 불필요한 의존성까지 제거해주시면 좋을 것 같아요.

popotec committed:
chore: 미사용 참조 삭제

popotec committed:
refactor: Question의 List<Answer> 일급콜렉션으로 래핑

popotec committed:
chore: 변수명 수정

popotec committed:
style: 코드 포맷팅

popotec committed:
style: 코드 포맷팅

popotec committed:
refactor: 미사용 import 삭제 및 메서드명 변경

popotec committed:
refactor: DeleteHistories 일급콜렉션 정의 및 DeleteHistory 생성 방법 변경

1. List<DeleteHistory> 형태의 자료를 DeleteHistories로 래핑하여 삭제 정보 생성 로직도 위임
2. DeleteHistory 생성자를 private으로 변경하고 정적 팩토리 메서드로 생성하도록 변경
- DeleteHistory를 클라이언트가 직접 ConetentType과 id를 넘기지 않고 Question, Answer 객체만 넘기도록 변경 => DeleteHistory 생성 로직이 변경되도 클라이언트의 영향 축소
3. DeleteHistory 생성시 createDate를 직접 넣지 않고 생성자에서 처리하도록 변경
4. 정적 팩토리 메서드에서는 임의로 DeleteHistory를 생성할때 삭제되지 않은 컨텐츠를 추가하지 않도록 validation 체크
5. Question의 삭제 로직과 DeleteHistories 생성 간 연계성을 위해 Question.delete의 작업 완료 결과로 DeleteHistories를 생성하여 반환하도록 변경

popotec commented:
동환님 조언에 따라 List<Answer> , List<DeleteHistory> 와 같이 사용하던 것을 일급컬렉션으로 래핑하니 자연스럽게 역할을 위임할수 있게 된 것 같습니다!😊
보다 도메인 주도적인 설계가 된 것 같아 이전 코드보다 훨씬 만족스럽습니다. 
좋은 방향성 알려주셔서 감사합니다.👍

na-mul commented:
앗 요거 저도 놓친 부분이긴 한데, 

Entity 각각 equals 메소드에서

```
		if (o instanceof User) {
			return false;
		}
```

위와 같은 로직이 있으면 수정 부탁드립니다. 
User 클래스의 equals 메소드 일부를 가져와봤는데, if 안의 조건문을 반대로 할 필요가 있어보입니다.

na-mul commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,43 @@
+@Embeddable
+public class Answers {
+
+	@OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
+	private List<Answer> answers = new ArrayList<>();
```
`cascade = CascadeType.ALL`는 신중히 사용하여야 합니다. 

이 옵션에 의해 `Question`이 삭제될 때, 자식 `Answer`도 함께 삭제되게 됩니다.
만일 이 옵션이 없다면, 외래 키 제약조건에 의해 `Answer`가 등록된 `Question`은 삭제 시도가 에러가 발생하겠지요.
그런 에러가 썩 달갑지는 않지만 여하튼 데이터가 의도치않게 삭제되지는 않기 때문에, 
특별히 의도하지 않았으면 `cascade = CascadeType.ALL`은 빼는 편이 일반적일 것입니다.

그리고 정말 `cascade = CascadeType.ALL`를 넣을 예정이시면,
`orphanRemoval = true` 옵션도 함께 추가해보는 것도 고려해볼만 합니다.
그렇게되면 answers 리스트에서 `Answer`를 remove하는 것만으로도 실제 데이터 삭제가 이뤄지지요. 
`Question`만으로도 `Answer`의 생명주기를 모두 관리할 수 있게 됩니다.

물론 이번 과제에서는 실제 데이터가 삭제될 일이 없고 delete 플래그만 true로 설정하기에 큰 의미 없는 상황이기는 합니다.

이것저것 신경쓰지 않고 가장 쓰기 편한 cascade설정은 `cascade = CascadeType.PERSIST`가 될 것 같아요.
이거라면 `QuestionRepository` 에서 `Question`을 save할 때, `Answer`까지 자동으로 save되게 됩니다. 
그 이상의 영속성 전이 동작은 없으니 괜찮을 것 같아요.

popotec replied:
안녕하세요 동환님🙂
```cascade = CascadeType.ALL``` 옵션에 대해 깊이 생각해보지 못하고 적용했던것 같습니다.

삭제 시도시, 사용자가 question이나 answer의 등록자인지도 체크하는 것이 요구사항이기 때문에 Question이 삭제된다고 Answer이 연쇄적으로 삭제되는 것은 좋은 방향이 아닌 것 같습니다.(실제 삭제가 아니고 deleted를 사용하기에 더더욱 부적합한 방향인 것 같습니다.😅)

위 요구사항이 아니어도 실무에서 예기치 못한 삭제를 일으킬수 있는 요소는 제거하고 영향도를 최소화 하는게 맞는 것 같습니다.
DB 데이터를 직접 만지다 question을 삭제하면 실수로 다 지워질 위험도 있을 것 같구요. 😅

비즈니스 흐름이 Question과 Answer를 한 사용자가 동시에 등록하진 않을 것 같아서 cascade 옵션 자체를 제거하였습니다.

중요한 포인트 짚어주셔서 감사합니다.👍

na-mul reviewed (changes requested):
안녕하세요 지희님,
코멘트 드린대로 잘 수정해주셨네요. 정말 감사합니다!
이대로 approve하고 미션을 종료해도 괜찮지만,
그래도 조금 더 소통하고자 하는 마음에 코멘트와 함께 Request changes를 남겨봅니다.🙇‍♂️ 
부담갖지 마시고 간단히 확인 후 다시 Request 올려주세요~ 👍

popotec committed:
fix: equals 인스턴스 타입 체크 구문 오류 수정

popotec committed:
fix: 부적절한 cascade 옵션 제거

na-mul reviewed (approved):
안녕하세요, 지희님.
코멘트 드린 것에 대하여 잘 반영된 것 확인하였습니다. 👍
짧지만 중요한 JPA 미션 고생 많으셨습니다.
성장의 계기가 되는 미션이었길 바라는 마음입니다.
남은 다른 미션에서도 더 많은 것을 배우고, 또 수료까지 잘 이루실 수 있기를 응원할게요!🙏
추가적인 DM이나 코멘트도 언제든 가능하니 부담없이 호출해주세요~ 🙇‍♂️

popotec commented:
> 안녕하세요, 지희님. 코멘트 드린 것에 대하여 잘 반영된 것 확인하였습니다. 👍 짧지만 중요한 JPA 미션 고생 많으셨습니다. 성장의 계기가 되는 미션이었길 바라는 마음입니다. 남은 다른 미션에서도 더 많은 것을 배우고, 또 수료까지 잘 이루실 수 있기를 응원할게요!🙏 추가적인 DM이나 코멘트도 언제든 가능하니 부담없이 호출해주세요~ 🙇‍♂️

안녕하세요, 동환님. 
JPA를 실무에선 사용해보지 못하고 동영상 강의를 보며 공부만 해보았었는데요, 미션을 진행하면서 공부 했던 것을 적용해보는 과정에서 명확하게 이해하지 못했던 개념들이 이제는 조금 선명해진 것 같습니다. 여기에는 부적절한 사용에 대한 근본적인 원인을 제고해보게 하고 더 나은 방향을 제안해주신 동환님 덕분이라고 꼭 말씀드리고 싶습니다. 탄탄한 기초가 필요하단 생각이 들어 지금은 탄력받아서 김영한님 JPA 책 구매해서 보는 중입니다.🙂
바쁘신 일상이 있음에도 적극적으로 도와주셔서 감사했습니다.🙇‍♀️
